Added examples for MAX31856 thermocouple sensor and waterproof ultrasonic distance sensor

RBerliner opened the pull request:
I have added code to support the Adafruit MAX31856 breakout board for a thermocouple sensor. I use it to monitor the engine temperature.

I have also added code for an ultrasonic waterproof distance sensor that I use for a fresh water tank monitor. 

I have implemented suggestions provided by Brian Smith

The ultrasonic distance sensor has been implemented with one class (UltrasonicSens) rather than adding an additional class (UltrasonicSensValue)

ba58smith commented on src/sensors/max31856TC_input.h:
```diff
@@ -0,0 +1,28 @@
+#define SPI_MISO_PIN 12
+#define SPI_CLK_PIN 14
+#define DRDY_PIN 5
+
```
Since SensESP is a library, we don't want the end user to have to modify the source files if we can help it. In this case, we can definitely help it. I will again refer to the code for the ina219 - it first creates a class that represents the actual sensor `INA219::INA219`. That class has only the constructor - which sets all of the low-level stuff necessary to get the physical sensor working (the hex address and the calibration settings) - and a pointer to an instance of an Adafruit_INA219.
Then, the value of the sensor is read by the class INA219Value. It has a pointer to the sensor class (pINA219), the read_delay, and all the details needed to read that value, and configure the sensor with values that can be changed on the fly.
Please follow that same format here: make a new class (like "MAX31856"), the constructor of which has all five of the pins, the thermocouple type, and the conversion mode. Those seem to be the things that are necessary to "start up" the sensor, so that it's ready to start reading values, correct? It will have a data member that is a pointer to an Adafruit_MAX31856, something like `pAdafruitMAX31856`.
Then, modify your `MAX31856TCInput` class to follow the pattern of the `INA219Value` class (and maybe change the name to `MAX31856Value` to be consistent with many of the other SensESP sensors). It will have a pointer to the new class above (`MAX31856* pMAX31856`), and all the code needed to read the sensor, and configure the things that can be changed on the fly.

ba58smith commented on examples/max31856TC_example.cpp:
```diff
@@ -0,0 +1,69 @@
+ 
+  const float multiplier = 1.0;
+  const float offset = 273.16;
+
```
Since `multiplier` and `offset` will never be changed, you don't need to make them configurable. There's no harm in doing so, but no value, either. To make them not configurable, simply remove `linear_config_path` from this file - the definition of it above, and the use of it below. This is totally up to you - if you can think of a situation in which someone might want to configure them, leave it as-is.

ba58smith commented on src/sensors/ultrasonic_input.h:
```diff
@@ -0,0 +1,25 @@
+
+#define TRIGGER_PIN 15
+#define INPUT_PIN 14
+
```
Please move these two pins to the constructor of this class. That will allow the end user to select their pins in main.cpp, rather than having to modify a source file. Since they always need to be specified (everyone is going to wire their project differently), make them the first two parameters in the constructor, followed by `read_delay` and the config path. That's so you have to specify the read delay and the config path if you want to - if you DON'T want to modify the read delay, you can pass in only the two pins parameters.

ba58smith commented on src/sensors/ultrasonic_input.cpp:
```diff
@@ -0,0 +1,60 @@
+  pulseWidth = pulseIn(INPUT_PIN, HIGH);
+  return pulseWidth;
+}
+
```
This arrangement works, but `update()` doesn't do much. You could simplify this by making `update()` do everything that `utrasonicRead()` does, then call `notify()`.
(I know - you copied this from some other sensor code that has a separate function for enable, reading, and updating, which isn't wrong - but IMO, it's not necessary, and certainly not simple. I know that I also told you, in the MAX31856 code, to combine all three of those into `enable()`. But in this case, since there's quite a bit of logic in `ultrasonicRead(), it think it's good to keep that mostly separate - except for adding the call to `notify()` to the end of that logic. So you'll have two functions instead of three - one being very "boilerplate", and one with the sensor read logic.)

ba58smith commented on src/sensesp_app.cpp:
```diff
@@ -9,6 +9,7 @@
 #include "sensors/digital_input.h"
 #include "sensors/system_info.h"
+#include "sensors/ultrasonic_input.h"
 #include "net/discovery.h"
```
I don't think you need the #include for ultrasonic_input.h here - it's in main.cpp. (I've not looked at this before - I don't know why analog_input.h and digital_input.h are included here, either.)

Please try your code without including ultrasonic_input.h here, and if it works, remove it. Then, if you want, please try to compile the analog input example without including analog_input.h here, and the RPM counter example with including digital_input.h here. If they all work, please remove all three of those includes from here. Thanks!

ba58smith replied:
Did you try it without `#include "sensors/ultrasonic_input.h"`? If you did, and it won't compile, can you show me the error? I see no reason this #include needs to be in sensesp_app.h.

ba58smith commented on examples/README_MAX31856TC.md:
```diff
@@ -0,0 +1,84 @@
+## Software
+
+The Thermocouple engine temperature sensor is implemented in software with two files src/sensors/max31856TC_input.h and src/sensors/max31856_input.cpp. The MAX31856 breakout board communicates with the Huzzah through software SPI and if you want to use pins other than SPI_CS_PIN 15, SPI_MOSI_PIN 13, SPI_MISO_PIN 12 and SPI_CLK_PIN 14 or a thermocouple other than Type-K, you will need to make the  ...
+
```
Please modify the above after you make the changes I've requested below (changes that make it so the end user doesn't have to modify the .h file.).

ba58smith commented on examples/README_ultrasonic_DVP.md:
```diff
@@ -0,0 +1,48 @@
+In practice I found that with T1 = 0.1 ms, T2 ~ 15 ms. The distance (in cm) to the reflecting object is given by 2*distance = pulse_width/speed_of_sound with the pulse_width (T3) in micro-sec and the the 2*speed of sound = 58 micro-sec/cm.
+
+If you use an Adafruit Feather Huzzah ESP8266 and the same wiring as above, you can use SensESP as a library. If  you intend to change wiring, you will have to adjust the definitions in src/sensors/ultrasonic_input.h and recompile SensESP. 
+
```
Please modify the above after you make the changes I've requested below (changes that make it so the end user doesn't have to modify the .h file.).

ba58smith commented on examples/ultrasonic_DVP_example.cpp:
```diff
@@ -0,0 +1,71 @@
+  // run-time configuration.
+  const char* ultrasonic_in_config_path = "/freshWaterTank_starboard/ultrasonic_in";
+  const char* linear_config_path = "/freshWaterTank_starboard/linear";
+  
```
See my comment about config paths in your other example file.

RBerliner commented:
Dear Brian:

First thank you for the constructive criticism. My hat is off to the
authors of this code - it is above my coding pay-grade and it has been fun
(though occasionally frustrating) to implement changes.
I am studying your comments and will try to implement them.

I completely understand your point about including the pins and so on in
the constructor in main.cpp. I, too, thought the need to rebuild SensESP
for connection/TC-type  was awkward. What I do not currently understand is
how to build SensESP for testing.

When I did it, I changed lib_deps line https://github/SignalK/SensESP.git
to c:/Path/to/my/repo/SensESP and added all of the other libraries needed
to lib_deps. This is basically the list that is in library.json. Then I
could compile and create a library. .Is this how you do it?

Another thing I notice is that the README talks about using ESP32. When I
first tried to do this - a couple of months ago, using an ESP32 device
caused SEensESP compile to fail. Does this now work?

I will implement the changes and may ask further questions. Working alone
is difficult.

Ron

On Sat, Jun 27, 2020 at 10:51 AM Brian Smith <notifications@github.com>
wrote:

> *@ba58smith* requested changes on this pull request.
> ------------------------------
>
> In examples/README_MAX31856TC.md
> <https://github.com/SignalK/SensESP/pull/135#discussion_r446523250>:
>
> > +In pactice, the MAX31856 is screewed to the top of the Huzzah on a Feather prototyping "Featherwing" with nylon screws and short nylon standoffs. Electrical connections have been made with wire-wrap. It is easier to alter if later if you want to alter the wiring.
> +
> +![picture 5](images/a773526aa785ac92d575d192a85b010faae1391ef46b2f8bae0288978744f173.png)
> +
> +![picture 1](images/b0c2954b815f65e47b3dd5c575bed96e8eb4311e7067c1b201d160dbcff6e1c6.png)
> +
> +The Max31856 thermocouple amplifier will not work with a grounded thermocouple tip. I have placed two layers of shrink tubing around the TC tip and then clamped it to the engine between two washers. This seems to work but the metal washers might be replaced with nylon which melts about 220 C - higher than needed for the engine temperature applicaiton.
> +
> +## Software
> +
> +The Thermocouple engine temperature sensor is implemented in software with two files src/sensors/max31856TC_input.h and src/sensors/max31856_input.cpp. The MAX31856 breakout board communicates with the Huzzah through software SPI and if you want to use pins other than SPI_CS_PIN 15, SPI_MOSI_PIN 13, SPI_MISO_PIN 12 and SPI_CLK_PIN 14 or a thermocouple other than Type-K, you will need to make the appropriate adjustments in max31856TC_input.h and recompile SensESP.
> +
> +To make changes to SensESP, fork the SensESP repo and then clone it to your local PC.
> +
> +To implement thermocouple engine temperature sensor, create a new PlatformIO project, select the Adafruit HUZZAH ESP8266 as the board, select Arduino as the framework, check the box "Use default location" and copy the contents of the file max31856TC_example.cpp into src/main.cpp.
> +
>
> This isn't how all those libraries should be included in the project. (I
> thought it was done this way, too, so don't feel bad!) The only thing you
> need in the lib_deps section is the reference to SensESP. All the other
> libraries, which are part of SensESP, belong in
> /PlatformIO/Projects/SensESPTest/.pio/libdeps/d1_mini/SensESP/library.json.
> If you do it like below, all of those libraries are included in the
> compiled program. If you put them all in library.json, PlatformIO will link
> in only the ones it needs for the current project.
>
> I don't see an additional library added for the MAX31856, and I think
> that's because it uses 1-Wire technology, which is already included,
> correct?
> ------------------------------
>
> In examples/max31856TC_example.cpp
> <https://github.com/SignalK/SensESP/pull/135#discussion_r446524796>:
>
> > +  // this blank would indicate this particular sensor (or transform) does not
> +  // broadcast SignalK data.
> +  const char* sk_path = "propulsion.Main_Engine.temperature";
> +
> +
> +  // The "Configuration path" is combined with "/config" to formulate a URL
> +  // used by the RESTful API for retrieving or setting configuration data.
> +  // It is ALSO used to specify a path to the SPIFFS file system
> +  // where configuration data is saved on the MCU board. It should
> +  // ALWAYS start with a forward slash if specified. If left blank,
> +  // that indicates this sensor or transform does not have any
> +  // configuration to save, or that you're not interested in doing
> +  // run-time configuration.
> +  const char* temperature_in_config_path = "/propulsion/Main_Engine/temperature_in";
> +  const char* linear_config_path = "/Main_Engine/temperature_in/linear";
> +
>
> I don't know if these two config paths will work. The Config UI will (I
> think) make a layout like this:
> propulsion -> Main_Engine -> temperature_in and another one like this: Main_Engine
> -> temperature_in -> linear. Note that temperature_in is both a "level"
> (between "Main_Engine" and "linear"), and an "endpoint" (after
> "Main_Engine").
> Based on how you're using these two config paths in the code below, I
> think what you want is this:
>
> const char* temperature_in_config_path = "/propulsion/Main_Engine/temperature_in/read_delay";
> const char* linear_config_path = "/propulsion/Main_Engine/temperature_in/linear";
>
> Notice that I added propulsion/ to the beginning of the second one -
> that's for a nicer looking UI. You could also delete propulsion/ from
> both of them, and, unless you're going to have other Main_Engine
> measurements on this ESP, you could REALLY shorten these, to simply
> "/read_delay" and "/linear". The more levels you have, the more clicking
> you have to do when using the Config UI.
> ------------------------------
>
> In library.json
> <https://github.com/SignalK/SensESP/pull/135#discussion_r446525745>:
>
> > @@ -1,95 +1,95 @@
>  {
>
> In future Pull Requests, please don't do massive reformatting like this.
> It makes it very hard for anyone to identify the actual changes that you've
> made, since EVERYTHING changes.
> I think the only REAL change you made was adding the Adafruit MAX31856
> library, correct?
> ------------------------------
>
> In platformio.ini
> <https://github.com/SignalK/SensESP/pull/135#discussion_r446525920>:
>
> > @@ -8,9 +8,9 @@
>  ; Please visit documentation for the other options and examples
>  ; https://docs.platformio.org/page/projectconf.html
>
> Please remove this file from the Pull Request. Since it's going to be
> different for every user's hardware, and is not part of the SensESP library
> code, just take it out.
> ------------------------------
>
> In src/sensors/max31856TC_input.cpp
> <https://github.com/SignalK/SensESP/pull/135#discussion_r446526215>:
>
> > @@ -0,0 +1,64 @@
> +#include "max31856TC_input.h"
> +
> +#include "Arduino.h"
> +#include "sensesp.h"
> +#include "Adafruit_MAX31856.h"
> +
>
> #include "Adafruit_MAX31856.h" should be in your .h file (at line 13),
> not here in the .cpp file.
> ------------------------------
>
> In src/sensors/max31856TC_input.cpp
> <https://github.com/SignalK/SensESP/pull/135#discussion_r446527729>:
>
> > +#include "Adafruit_MAX31856.h"
> +
> +Adafruit_MAX31856 max31856TC = Adafruit_MAX31856(SPI_CS_PIN, SPI_MOSI_PIN, SPI_MISO_PIN, SPI_CLK_PIN);
> +
> +
> +MAX31856TCInput::MAX31856TCInput(uint read_delay, String config_path)
> +    : NumericSensor(config_path), read_delay{read_delay} {
> +  className = "MAX31856TCInput";
> +  if (!max31856TC.begin()) {
> +	  while(1) delay(10);
> +  }
> +  max31856TC.setThermocoupleType(MAX31856_TCTYPE_K);
> +  max31856TC.setConversionMode(MAX31856_CONTINUOUS);
> +  load_configuration();
> +}
> +
>
> You can simplify this a lot, by combining update(), tcRead(), and enable().
> Look at how enable() is implemented in ina219.cpp: it starts with the
> app.onRepeat line, then does the code necessary to read the sensor and
> update the value of output, then calls notify(). You can do the same by
> putting the code of tcRead() directly into enable(), then call notify().
> This may not save any clock cycles, but it will be much easier to read.
> ------------------------------
>
> In src/sensors/max31856TC_input.h
> <https://github.com/SignalK/SensESP/pull/135#discussion_r446529926>:
>
> > @@ -0,0 +1,28 @@
> +#ifndef _MAX31856TC_input_H_
> +#define _MAX31856TC_input_H_
> +
> +#define SPI_CS_PIN 15
> +#define SPI_MOSI_PIN 13
> +#define SPI_MISO_PIN 12
> +#define SPI_CLK_PIN 14
> +#define DRDY_PIN 5
> +
>
> Since SensESP is a library, we don't want the end user to have to modify
> the source files if we can help it. In this case, we can definitely help
> it. I will again refer to the code for the ina219 - it first creates a
> class that represents the actual sensor INA219::INA219. That class has
> only the constructor - which sets all of the low-level stuff necessary to
> get the physical sensor working (the hex address and the calibration
> settings) - and a pointer to an instance of an Adafruit_INA219.
> Then, the value of the sensor is read by the class INA219Value. It has a
> pointer to the sensor class (pINA219), the read_delay, and all the details
> needed to read that value, and configure the sensor with values that can be
> changed on the fly.
> Please follow that same format here: make a new class (like "MAX31856"),
> the constructor of which has all five of the pins, the thermocouple type,
> and the conversion mode. Those seem to be the things that are necessary to
> "start up" the sensor, so that it's ready to start reading values, correct?
> It will have a data member that is a pointer to an Adafruit_MAX31856,
> something like pAdafruitMAX31856.
> Then, modify your MAX31856TCInput class to follow the pattern of the
> INA219Value class (and maybe change the name to MAX31856Value to be
> consistent with many of the other SensESP sensors). It will have a pointer
> to the new class above (MAX31856* pMAX31856), and all the code needed to
> read the sensor, and configure the things that can be changed on the fly.
> ------------------------------
>
> In examples/max31856TC_example.cpp
> <https://github.com/SignalK/SensESP/pull/135#discussion_r446530205>:
>
> > +  // It is an ultrasonic distance sensor that sends out an acoustical pulse in response
> +  // to a 100 micro-sec trigger pulse from the ESP. The return acoustical pulse width
> +  // can be converted to a distance by the formula 2*distance = pulse_width/speed_of_sound
> +  // With pulse_width om ,icro-sec and distance in cm, 2*speed_of_sound = 58
> +  // The sensor is mounted at the top of a water tank that is 25 cm deep.
> +  uint read_delay = 500;
> +
> +  auto* pTemperatureInput = new MAX31856TCInput(read_delay, temperature_in_config_path);
> +
> +  // A Linear transform takes its input, multiplies it by the multiplier, then adds the offset,
> +  // to calculate its output. The MAX31856TC produces temperatures in degrees Celcius. We need to change
> +  // them to Kelvin for compatibility with SignalK.
> +
> +  const float multiplier = 1.0;
> +  const float offset = 273.16;
> +
>
> Since multiplier and offset will never be changed, you don't need to make
> them configurable. There's no harm in doing so, but no value, either. To
> make them not configurable, simply remove linear_config_path from this
> file - the definition of it above, and the use of it below. This is totally
> up to you - if you can think of a situation in which someone might want to
> configure them, leave it as-is.
> ------------------------------
>
> In src/sensors/ultrasonic_input.h
> <https://github.com/SignalK/SensESP/pull/135#discussion_r446531208>:
>
> > @@ -0,0 +1,25 @@
> +#ifndef _ultrasonic_input_H_
> +#define _ultrasonic_input_H_
> +
> +#define TRIGGER_PIN 15
> +#define INPUT_PIN 14
> +
>
> Please move these two pins to the constructor of this class. That will
> allow the end user to select their pins in main.cpp, rather than having to
> modify a source file. Since they always need to be specified (everyone is
> going to wire their project differently), make them the first two
> parameters in the constructor, followed by read_delay and the config
> path. That's so you have to specify the read delay and the config path if
> you want to - if you DON'T want to modify the read delay, you can pass in
> only the two pins parameters.
> ------------------------------
>
> In src/sensors/ultrasonic_input.cpp
> <https://github.com/SignalK/SensESP/pull/135#discussion_r446532193>:
>
> > +void UltrasonicInput::update() {
> +  output = ultrasonicRead();
> +  this->notify();
> +}
> +
> +int UltrasonicInput::ultrasonicRead() {
> +  digitalWrite(TRIGGER_PIN, HIGH);
> +  long lastTime = micros();
> +  while (micros() - lastTime < 100) {
> +    yield();
> +  }
> +  digitalWrite(TRIGGER_PIN, LOW);
> +  pulseWidth = pulseIn(INPUT_PIN, HIGH);
> +  return pulseWidth;
> +}
> +
>
> This arrangement works, but update() doesn't do much. You could simplify
> this by making update() do everything that utrasonicRead() does, then
> call notify().
> (I know - you copied this from some other sensor code that has a separate
> function for enable, reading, and updating, which isn't wrong - but IMO,
> it's not necessary, and certainly not simple. I know that I also told you,
> in the MAX31856 code, to combine all three of those into enable(). But in
> this case, since there's quite a bit of logic in ultrasonicRead(), it
> think it's good to keep that mostly separate - except for adding the call
> to notify()` to the end of that logic. So you'll have two functions
> instead of three - one being very "boilerplate", and one with the sensor
> read logic.)
> ------------------------------
>
> In src/sensesp_app.cpp
> <https://github.com/SignalK/SensESP/pull/135#discussion_r446533560>:
>
> > @@ -9,6 +9,7 @@
>  #include "sensors/analog_input.h"
>  #include "sensors/digital_input.h"
>  #include "sensors/system_info.h"
> +#include "sensors/ultrasonic_input.h"
>  #include "net/discovery.h"
>
> I don't think you need the #include for ultrasonic_input.h here - it's in
> main.cpp. (I've not looked at this before - I don't know why analog_input.h
> and digital_input.h are included here, either.)
>
> Please try your code without including ultrasonic_input.h here, and if it
> works, remove it. Then, if you want, please try to compile the analog input
> example without including analog_input.h here, and the RPM counter example
> with including digital_input.h here. If they all work, please remove all
> three of those includes from here. Thanks!
> ------------------------------
>
> In examples/README_MAX31856TC.md
> <https://github.com/SignalK/SensESP/pull/135#discussion_r446533768>:
>
> > +The connections for the Huzzah and MAX31856 are shown in the figure below.
> +
> +![picture 1](images/Huzzah-Max31856_bb.png)
> +
> +In pactice, the MAX31856 is screewed to the top of the Huzzah on a Feather prototyping "Featherwing" with nylon screws and short nylon standoffs. Electrical connections have been made with wire-wrap. It is easier to alter if later if you want to alter the wiring.
> +
> +![picture 5](images/a773526aa785ac92d575d192a85b010faae1391ef46b2f8bae0288978744f173.png)
> +
> +![picture 1](images/b0c2954b815f65e47b3dd5c575bed96e8eb4311e7067c1b201d160dbcff6e1c6.png)
> +
> +The Max31856 thermocouple amplifier will not work with a grounded thermocouple tip. I have placed two layers of shrink tubing around the TC tip and then clamped it to the engine between two washers. This seems to work but the metal washers might be replaced with nylon which melts about 220 C - higher than needed for the engine temperature applicaiton.
> +
> +## Software
> +
> +The Thermocouple engine temperature sensor is implemented in software with two files src/sensors/max31856TC_input.h and src/sensors/max31856_input.cpp. The MAX31856 breakout board communicates with the Huzzah through software SPI and if you want to use pins other than SPI_CS_PIN 15, SPI_MOSI_PIN 13, SPI_MISO_PIN 12 and SPI_CLK_PIN 14 or a thermocouple other than Type-K, you will need to make the appropriate adjustments in max31856TC_input.h and recompile SensESP.
> +
>
> Please modify the above after you make the changes I've requested below
> (changes that make it so the end user doesn't have to modify the .h file.).
> ------------------------------
>
> In examples/README_ultrasonic_DVP.md
> <https://github.com/SignalK/SensESP/pull/135#discussion_r446533848>:
>
> > +| Output   | IO14   |  White       |
> +| Gnd      |  Gnd   | Black        |
> +
> +
> +![picture 1](images/Huzzah-UltrasonicDYP_bb_small.png)
> +
> +## Software
> +
> +The ultrawsonic distance sensor is implemented in software with the two files SensESP/src/sensors/ultrasonic_input.cpp and SensESP/src/sensors/ultrasonic_input.h. According to the manufacturer, the receipt of a Trigger pulse 100 microsec wide causes the sensor to emit a pulse whose width is proportional to 2X the travel time between the sensor and the reflecting surface.
> +
> +![picture 2](images/PWM-Waveform.png)
> +
> +In practice I found that with T1 = 0.1 ms, T2 ~ 15 ms. The distance (in cm) to the reflecting object is given by 2*distance = pulse_width/speed_of_sound with the pulse_width (T3) in micro-sec and the the 2*speed of sound = 58 micro-sec/cm.
> +
> +If you use an Adafruit Feather Huzzah ESP8266 and the same wiring as above, you can use SensESP as a library. If  you intend to change wiring, you will have to adjust the definitions in src/sensors/ultrasonic_input.h and recompile SensESP.
> +
>
> Please modify the above after you make the changes I've requested below
> (changes that make it so the end user doesn't have to modify the .h file.).
> ------------------------------
>
> In examples/ultrasonic_DVP_example.cpp
> <https://github.com/SignalK/SensESP/pull/135#discussion_r446534021>:
>
> > +  // this blank would indicate this particular sensor (or transform) does not
> +  // broadcast SignalK data.
> +  const char* sk_path = "tanks.freshWater.starboard.currentLevel";
> +
> +
> +  // The "Configuration path" is combined with "/config" to formulate a URL
> +  // used by the RESTful API for retrieving or setting configuration data.
> +  // It is ALSO used to specify a path to the SPIFFS file system
> +  // where configuration data is saved on the MCU board. It should
> +  // ALWAYS start with a forward slash if specified. If left blank,
> +  // that indicates this sensor or transform does not have any
> +  // configuration to save, or that you're not interested in doing
> +  // run-time configuration.
> +  const char* ultrasonic_in_config_path = "/freshWaterTank_starboard/ultrasonic_in";
> +  const char* linear_config_path = "/freshWaterTank_starboard/linear";
> +
>
> See my comment about config paths in your other example file.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/SignalK/SensESP/pull/135#pullrequestreview-438696115>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4CSSMEBHG6LYEXGP5M4K3RYYBNXANCNFSM4OJZPRLA>
> .
>


-- 

R. Berliner
Instrumentation Associates
101 Echowood Ln
Cary, NC 27518

e-mail: rberliner@InstrumentationAssociates.com

www.InstrumentationAssociates.com

Tel     734-646-5318

ba58smith commented:
Ron, I'm with you, man - when I started working on SensESP, I was really struggling to find my way. Writing several sensor integrations, and using SensESP on my boat on three different ESP's, has really upped my confidence (and, I hope, my competence). 

Yes, that's how you work with SensESP locally - point the SensESP entry in `lib_deps` to the SensESP folder on your hard drive - the folder where you have GitHub set up. When you do a `git pull upstream master`, you're pulling all the changes from the SensESP repo online (that's "upstream") to your local repo, using the "master" branch. So when you build SensESP, that's where it needs to point. Mine is `file:///home/butch/Documents/GitHubProjects/SensESP`. When the PlatformIO SensESP project is built, it pulls files from that location to compile and build, and puts the working files in your PlatformIO/Projects/SensESPWhatever folder. So when you're working in PlatformIO, you're modifying and compiling and building those files. But when you're ready to do the Pull Request (or update it, as you'll need to do now), you have to copy those working files into your local repository files, in order for git to know that you've made changes. (There's a way to do this all from within PlatformIO, but I don't do it that way, as I don't 100% understand it.) 
Using git and GitHub have been a real learning experience for me, and I still have to follow a very detailed set of instructions every time I do anything. Attached is my "working with SensESP and GitHub" cheat sheet, if it helps you at all.
[How to set up Git and SensESP.txt](https://github.com/SignalK/SensESP/files/4841521/How.to.set.up.Git.and.SensESP.txt)

Finally - yes, the ESP32 is now supposed to work. Someone else created the PR to make it all work, and I've tested it as much as I can, on both the ESP8266 (Wemos D1 Mini), and the ESP32 (a generic one I have). You can give it a try if you like, but I wouldn't do that unless you're working with something you KNOW works with an ESP8266.

RBerliner commented:
Thanks Butch:

I understand (more or less) the git part although I have suffered more than
one git-tastrophy when I did things wrong. The part about pointing the
lib_deps in the SensESP development branch to the local repo I also
understand.

You did not address the libraries when trying to build SensESP. How do you
add the libraries - I have added them explicitly in lib_deps below the
pointer to the local SensESP for the development build. Once SensESP
compiles you can create another VSCode project to do the particular sensor
example and then use the pointer to the local repo as a library. Then you
do not need any of the other libraries in lib_deps. It gets them from the
SensESP library.

Is this how you do it?

On Sat, Jun 27, 2020 at 9:22 PM Brian Smith <notifications@github.com>
wrote:

> Ron, I'm with you, man - when I started working on SensESP, I was really
> struggling to find my way. Writing several sensor integrations, and using
> SensESP on my boat on three different ESP's, has really upped my confidence
> (and, I hope, my competence).
>
> Yes, that's how you work with SensESP locally - point the SensESP entry in
> lib_deps to the SensESP folder on your hard drive - the folder where you
> have GitHub set up. When you do a git pull upstream master, you're
> pulling all the changes from the SensESP repo online (that's "upstream") to
> your local repo, using the "master" branch. So when you build SensESP,
> that's where it needs to point. Mine is
> file:///home/butch/Documents/GitHubProjects/SensESP. When the PlatformIO
> SensESP project is built, it pulls files from that location to compile and
> build, and puts the working files in your
> PlatformIO/Projects/SensESPWhatever folder. So when you're working in
> PlatformIO, you're modifying and compiling and building those files. But
> when you're ready to do the Pull Request (or update it, as you'll need to
> do now), you have to copy those working files into your local repository
> files, in order for git to know that you've made changes. (There's a way to
> do this all from within PlatformIO, but I don't do it that way, as I don't
> 100% understand it.)
> Using git and GitHub have been a real learning experience for me, and I
> still have to follow a very detailed set of instructions every time I do
> anything. Attached is my "working with SensESP and GitHub" cheat sheet, if
> it helps you at all.
> How to set up Git and SensESP.txt
> <https://github.com/SignalK/SensESP/files/4841521/How.to.set.up.Git.and.SensESP.txt>
>
> Finally - yes, the ESP32 is now supposed to work. Someone else created the
> PR to make it all work, and I've tested it as much as I can, on both the
> ESP8266 (Wemos D1 Mini), and the ESP32 (a generic one I have). You can give
> it a try if you like, but I wouldn't do that unless you're working with
> something you KNOW works with an ESP8266.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/SignalK/SensESP/pull/135#issuecomment-650663474>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4CSSK7GFPDWQ4WUKE6S2DRY2LPDANCNFSM4OJZPRLA>
> .
>


-- 

R. Berliner
Instrumentation Associates
101 Echowood Ln
Cary, NC 27518

e-mail: rberliner@InstrumentationAssociates.com

www.InstrumentationAssociates.com

Tel     734-646-5318

ba58smith commented:
Ron, I don't think you have to add the libraries (like for the MAX31856) to your local platformio.ini at all, ever. It should have ONLY the line that points to your local copy of SensESP code (as described above). When you modify the library.json file, modify it in the location that's pointed to by your local platformio.ini. In my case, I would modify the library.json file that's in the /home/butch/Documents/GitHubProjects/SensESP folder. Then, delete everything the "working" SensESP folder: /home/butch/Documents/PlatformIO/Projects/SensESPTest/.pio/libdeps/d1_mini/SensESP. Now, PlatformIO has no SensESP files to work with, so it looks at the ones in your local repo, and, seeing a new library in library.json, it builds SensESP with that new library. I hope that makes sense - I can't think of a better way to describe it.

ba58smith commented on examples/thermocouple_temperature_sensor/max31856TC_example.cpp:
```diff
@@ -61,7 +71,7 @@ ReactESP app([] () {
-  pTemperatureInput -> connectTo(new Linear(multiplier, offset, linear_config_path))
+  pMAX31856TCvalue -> connectTo(new Linear(multiplier, offset, ""))
                -> connectTo(new SKOutputNumber(sk_path));
 
```
This is really minor - you don't need to change it, I just want to point it out. You don't need to pass `" "` to Linear(), since that is the default value for the config path. Therefore, `Linear(multiplier, offset)` does exactly the same as `Linear(multiplier, offset, "")`.

ba58smith commented on src/sensors/max31856TC_input.cpp:
```diff
@@ -1,41 +1,41 @@
 
-Adafruit_MAX31856 max31856TC = Adafruit_MAX31856(SPI_CS_PIN, SPI_MOSI_PIN, SPI_MISO_PIN, SPI_CLK_PIN);
+uint8_t dataRdy;
 
```
dataRdy is a Pin. If you want to define all the pins in the same place, you're correct to have and use it here in the MAX31856TC class, but it should be a private member of the class, not a global variable. However, since it's not used in MAX31856TC, it doesn't need to be in this class. It's used in MAX31856TCValue, so it should be a private member of that class. You can have it be a member of both classes, but to know the value of it in MAX31856TCValue, it has to be a parameter to the constructor anyway, so why both having it in MAX31856TC at all? Put it in the "Value" class, and have it as a parameter to its constructor. The end user will provide it when instantiating the "Value" class in main.cpp, rather than when instantiating the MAX31856TC class.
Make it the second parameter in the constructor, as it should always be provided. `read_delay` and `config_path` don't always need to be provided, so having them be last means the end user doesn't need to provide them in main.cpp if they're happy with the default value for `read_delay`. Make sense?

ba58smith commented on src/sensors/max31856TC_input.cpp:
```diff
@@ -1,41 +1,41 @@
   }
-  max31856TC.setThermocoupleType(MAX31856_TCTYPE_K);
-  max31856TC.setConversionMode(MAX31856_CONTINUOUS);
-  load_configuration();
```
Does your code work for any type of thermocouple other than "K"? If it works only with Type K, you can drop the type from the constructor, and hard-code `MAX31856_TCTYPE_K` below.
Same question about "conversion mode".

ba58smith commented on examples/ultrasonic_level_sensor/ultrasonic_DVP_example.cpp:
```diff
@@ -3,32 +3,33 @@
+
+#define TRIGGER_PIN 15
+#device INPUT_PIN 14
 
```
Did you mean for this to be `#device`? I'm not familiar with that preprocessor directive.

ba58smith commented on src/sensors/ultrasonic_input.h:
```diff
@@ -1,25 +1,25 @@
+    UltrasonicSensValue(UltrasonicSens* pUltrasonicSens, uint read_delay = 500, String config_path = "");
+    void enable() override final;
+    UltrasonicSens* pUltrasonicSens;
 
```
This works (having two separate classes), but in this case, you really don't need the complexity. I meant for you to simply add the two pins to the constructor to your original `UltrasonicInput` class, to pull them out of the .h file:
`UltrasonicInput(int8_t trigger_pin, int8_t input_pin, uint read_delay = 200, String config_path = " ");`
That makes the usage in main.cpp simpler - just one line to create the sensor.
The only reason I think this is different than the MAX 31856 code is the number of constructor parameters. If you did everything with one class for the MAX, the constructor would have a LOT of parameters. (5 pins, the type, the conversion, the read delay, and the config path). It didn't HAVE to be done the way I had you do it. The reason INA219 has to be done that way is that sensor outputs more than one type of data (temp and humidity, I think?), so you need each of them to have a pointer to the one physical sensor. In the MAX31856 case, and the ultrasonic case, there is only one output, so everything could be done with a single class.
It's up to you - for both of your new sensors, now that I think about it some more (sorry!): leave them both with two classes (the sensor and the value), change each of them to a single class (with all the pins, of course), or a mixture of those two approaches.

ba58smith commented on src/sensors/ultrasonic_input.cpp:
```diff
@@ -1,37 +1,40 @@
-  pinMode(INPUT_PIN, INPUT_PULLUP);
+uint triggerPin;
+uint inputPin;
+
```
The pins need to be data members of a class. If you combine UltrasonicSens and UltrasonicSensValue (see my comment below for details about that), they will be private members of that one class. If you keep these two separate classes, the pins need to be members of the class that needs them, which is UltrasonicSensValue. Setting the pin mode for them doesn't need to be done in the "sensor" class - move it to the "value" class. (However - this is another argument for combining these two classes into one.) Again - sorry for the confusion I may have caused in my first round of comments.

ba58smith commented on src/sensors/ultrasonic_input.cpp:
```diff
@@ -1,37 +1,40 @@
+  inputPin = input_pin;
   load_configuration();
 }
 
```
When you move the pins inside the class, the two lines above that assign their values will go away, replaced by additions to the constructor definition, like `read_delay{read_delay}`.

ba58smith commented:
Ron, I think I have reviewed all of your most recent changes and made comments. But I'm not sure. So please make whatever changes are needed for this last round, then commit them, and I'll do one final review. Thanks!

RBerliner commented:
Brian:

This has been a good education for me.

I will have these changes today or tomorrow.

Ron

On Wed, Jul 1, 2020 at 8:16 AM Brian Smith <notifications@github.com> wrote:

> Ron, I think I have reviewed all of your most recent changes and made
> comments. But I'm not sure. So please make whatever changes are needed for
> this last round, then commit them, and I'll do one final review. Thanks!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/SignalK/SensESP/pull/135#issuecomment-652383489>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4CSSLM7UBCCAFFVC4AOOTRZMSLNANCNFSM4OJZPRLA>
> .
>


-- 

R. Berliner
Instrumentation Associates
101 Echowood Ln
Cary, NC 27518

e-mail: rberliner@InstrumentationAssociates.com

www.InstrumentationAssociates.com

Tel     734-646-5318

RBerliner commented:
I think I have satisfied all of your requests.

ba58smith commented on examples/ultrasonic_level_sensor/ultrasonic_DVP_example.cpp:
```diff
@@ -65,10 +64,11 @@ ReactESP app([]() {
   const float multiplier = -0.00074948;
   const float offset = 1.08675;
+  float scale = 1.0;
 
```
By making `multiplier` and `offset` constants, will they be able to be modified at runtime from the Config UI? I guess you'll have to try it.

RBerliner commented:
Maybe not - I will try it. Good catch.

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail&utm_term=icon>
Virus-free.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail&utm_term=link>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

On Sat, Jul 4, 2020 at 10:42 AM Brian Smith <notifications@github.com>
wrote:

> *@ba58smith* commented on this pull request.
> ------------------------------
>
> In examples/ultrasonic_level_sensor/ultrasonic_DVP_example.cpp
> <https://github.com/SignalK/SensESP/pull/135#discussion_r449778650>:
>
> > @@ -65,10 +64,11 @@ ReactESP app([]() {
>    // empty = 116 * (-0.044347 / 58) +  1.08675 = 0
>    const float multiplier = -0.00074948;
>    const float offset = 1.08675;
> +  float scale = 1.0;
>
>
> By making multiplier and offset constants, will they be able to be
> modified at runtime from the Config UI? I guess you'll have to try it.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/SignalK/SensESP/pull/135#pullrequestreview-442613566>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4CSSI2LWCCRL7NAYPU6BDRZ45TRANCNFSM4OJZPRLA>
> .
>


-- 

R. Berliner
Instrumentation Associates
101 Echowood Ln
Cary, NC 27518

e-mail: rberliner@InstrumentationAssociates.com

www.InstrumentationAssociates.com

Tel     734-646-5318

ba58smith commented on src/sensors/ultrasonic_input.cpp:
```diff
@@ -15,13 +12,15 @@ UltrasonicSens::UltrasonicSens(int8_t trig_pin, int8_t input_pin, String config_
   load_configuration();
```
```
triggerPin = trig_pin;
inputPin = input_pin;
```
works, but the preferred approach (because it's "cleaner", I suppose) is to save the value of a function parameter the way you've done with `read_delay{read_delay]`. Like this:
` NumericSensor(config_path), triggerPin{trig_pin}, inputPin{input_pin}, read_delay{read_delay} {'

ba58smith commented:
Ron, in addition to the comments I just added, there's still one significant issue that's an "Unresolved conversation" - the one about the data ready pin variable. The comment has a lot of details, but here's the gist of it: the variable needs to be a private member of the "value" class, passed in by the constructor. Currently, it's a global variable (a C++ no-no).

Almost there!

RBerliner commented:
I checked changing multiplier and offset from the web config page - it
works fine.

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail&utm_term=icon>
Virus-free.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail&utm_term=link>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

On Sat, Jul 4, 2020 at 10:42 AM Brian Smith <notifications@github.com>
wrote:

> *@ba58smith* commented on this pull request.
> ------------------------------
>
> In examples/ultrasonic_level_sensor/ultrasonic_DVP_example.cpp
> <https://github.com/SignalK/SensESP/pull/135#discussion_r449778650>:
>
> > @@ -65,10 +64,11 @@ ReactESP app([]() {
>    // empty = 116 * (-0.044347 / 58) +  1.08675 = 0
>    const float multiplier = -0.00074948;
>    const float offset = 1.08675;
> +  float scale = 1.0;
>
>
> By making multiplier and offset constants, will they be able to be
> modified at runtime from the Config UI? I guess you'll have to try it.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/SignalK/SensESP/pull/135#pullrequestreview-442613566>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4CSSI2LWCCRL7NAYPU6BDRZ45TRANCNFSM4OJZPRLA>
> .
>


-- 

R. Berliner
Instrumentation Associates
101 Echowood Ln
Cary, NC 27518

e-mail: rberliner@InstrumentationAssociates.com

www.InstrumentationAssociates.com

Tel     734-646-5318

RBerliner commented:
I took it out and it compiles fine.
It was probably me flailing around trying to figure out why things were not
compiling. I stuck it in - it didn't fix things and I neglected to take it
out.

Ron

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail&utm_term=icon>
Virus-free.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail&utm_term=link>
<#m_8464705016784498050_DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

On Sat, Jul 4, 2020 at 11:00 AM Brian Smith <notifications@github.com>
wrote:

> *@ba58smith* commented on this pull request.
> ------------------------------
>
> In src/sensesp_app.cpp
> <https://github.com/SignalK/SensESP/pull/135#discussion_r449780021>:
>
> > @@ -9,6 +9,7 @@
>  #include "sensors/analog_input.h"
>  #include "sensors/digital_input.h"
>  #include "sensors/system_info.h"
> +#include "sensors/ultrasonic_input.h"
>  #include "net/discovery.h"
>
> Did you try it without #include "sensors/ultrasonic_input.h"? If you did,
> and it won't compile, can you show me the error? I see no reason this
> #include needs to be in sensesp_app.h.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/SignalK/SensESP/pull/135#discussion_r449780021>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4CSSOX7AWJARPSFXGYJ5DRZ47X5ANCNFSM4OJZPRLA>
> .
>


-- 

R. Berliner
Instrumentation Associates
101 Echowood Ln
Cary, NC 27518

e-mail: rberliner@InstrumentationAssociates.com

www.InstrumentationAssociates.com

Tel     734-646-5318

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail&utm_term=icon>
Virus-free.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail&utm_term=link>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

RBerliner commented:
Aloof this is useful but I now think the MAX31856TC class should include in
its argument list the drdy_pin and read_delay.

To my mind the structure with a single class (instead of a sensor class and
a sensor_value class is cleaner. If you have no objections, I am going to
rewrite things so that there is only one class...

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail&utm_term=icon>
Virus-free.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail&utm_term=link>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

On Sat, Jul 4, 2020 at 11:23 AM Brian Smith <notifications@github.com>
wrote:

> Ron, in addition to the comments I just added, there's still one
> significant issue that's an "Unresolved conversation" - the one about the
> data ready pin variable. The comment has a lot of details, but here's the
> gist of it: the variable needs to be a private member of the "value" class,
> passed in by the constructor. Currently, it's a global variable (a C++
> no-no).
>
> Almost there!
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/SignalK/SensESP/pull/135#issuecomment-653778711>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4CSSKVP32BGSRI4QYJTG3RZ5COXANCNFSM4OJZPRLA>
> .
>


-- 

R. Berliner
Instrumentation Associates
101 Echowood Ln
Cary, NC 27518

e-mail: rberliner@InstrumentationAssociates.com

www.InstrumentationAssociates.com

Tel     734-646-5318

RBerliner commented:
I think we have them all.

Thanks for all of the help.

RBerliner commented:
Got through all of the pull request comments to discover that I had forgotten to push my last commit.

I have pushed it now.

ba58smith commented:
Ron, I agree with your decision to have only one MAX31856 class, and again, my apologies for suggesting otherwise. I'm sure it was frustrating - I hope it was educational.
Everything looks good - I'm merging the PR (finally!).
A few comments for your next PR:

1. When replying to a particular review comment, please type your reply right there with the original comment. This makes it far easier for the reviewer to keep track of what has been said and done with respect to any particular comment. When the comments are added to the end of the whole conversation, it's much harder to know what has been addressed.

2. Please make each enhancement a separate PR. In this case, the MAX31856 and the ultrasonic sensor would each be in a separate PR. Again, this is to make the review process a LOT easier.

3. I already said this somewhere above, but it bears repeatiing: please don't do a massive reformat of the code (spacing and layout) as part of another PR. That makes it look like every line in a file has changed, making it VERY difficult to identify any actual changes to the code.

I look forward to your next PR!
